### PR TITLE
[Serializer] Skip test if `symfony/property-info` doesn't support `phpstan/phpdoc-parser:2.0` yet

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

The support for `phpstan/phpdoc-parser` 2.0 was introduced in PropertyInfo lately (cc @xabbuh). Unfortunately, Serializer cannot benefit from this in 5.4 because the highest version of PropertyInfo installable is 6.3 (see [the failing job](https://github.com/symfony/symfony/actions/runs/12014213386/job/33489484938)). Before running this particular test, we need to ensure PropertyInfo is installed with a version that supports `phpstan/phpdoc-parser` 2.0.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php [hunk 1]
```diff
@@ -753,6 +754,14 @@ public function testDoesntHaveIssuesWithUnionConstTypes()
             $this->markTestSkipped('phpstan/phpdoc-parser required for this test');
         }
 
+        $versionRange = InstalledVersions::getVersionRanges('symfony/property-info');
+
+        if (
+            '5.4.x-dev' !== $versionRange && version_compare($versionRange, '5.4.47', '<')
```
Comment (alexandre-daubois) [inline src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php:760]: version_compare() doesn't handle ".x-dev" versions well, thus the hard check

**Comment by wouterj** [comment]
Hi!

As far as I understand, this means that Serializer is broken for users when phpdoc-parser 2.0 is installed? Shouldn't we conflict with it in that case or add it as a dev dependency with a version range limit to 1.x?

**Comment by alexandre-daubois** [comment]
That was indeed what my first commit contained! Then we had a discussion with @xabbuh about it. In fact, the problem with doing this is that it blocks compatibility with version 2.0 in 5.4 (and other maintained branches), where support does work on the latest patch releases.

Indeed, the problem is that support is not available on unmaintained versions (6.3 for example). We thought it would be better to just skip the test rather than penalize maintained versions (Serializer installs PropertyInfo 6.3 in high deps). We agreed that if the dev installs versions that are no longer maintained, it's his/her responsibility.

I hope I've made myself clear, I struggled a bit summarizing the whole conversation 😄

File: src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php [hunk 2]
```diff
@@ -753,6 +754,14 @@ public function testDoesntHaveIssuesWithUnionConstTypes()
...
+            || version_compare($versionRange, '5.4.48', '>=') && '6.4.x-dev' !== $versionRange && version_compare($versionRange, '6.4.15', '<')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php]: not sure I get the constraint on 5.4.48, shouldn't this be on >= 6?
```suggestion
            || version_compare($versionRange, '6', '>=') && '6.4.x-dev' !== $versionRange && version_compare($versionRange, '6.4.15', '<')
        ) {
```

**Comment by nicolas-grekas** [comment]
Actually, the fix might be to remove that conflict in 6.4:
https://github.com/symfony/symfony/blame/403a9c3ff023ac2249d04cf8ce71fc57f9434781/src/Symfony/Component/PropertyInfo/composer.json#L40

**Comment by alexandre-daubois** [comment]
Indeed, this works pretty well! Let me update the PR accordingly. Thanks for the solution.

**Comment by alexandre-daubois** [comment]
~I tested it and yes, it actually seems to work. Do I update the PR accordingly, or should we close it in favor of yours?~

Nevermind 😄

**Comment by nicolas-grekas** [comment]
Replaced by https://github.com/symfony/symfony/pull/59008 of course.

**Pull request closed without merging**